File: gaml.grammar/src/gaml/grammar/transition/fix_with_pairs.py
#!/usr/bin/env python3
"""fix_with_pairs.py

Transforms ``with:`` argument lists in GAML source files from the old
bracket / double-colon pair syntax to the new parenthesis / single-colon
syntax.

Examples
--------
    with: [with_viz::true, step::1#s, name::"Simulation step=1s"]
        →  with: (with_viz:true, step:1#s, name:"Simulation step=1s")

Rules applied
-------------
1. The opening ``[`` immediately following ``with:`` (with optional
   whitespace) is replaced by ``(``.
2. The closing ``]`` that ends the list is replaced by ``)``.
3. Every ``::`` inside the list is replaced by ``:``.

The script handles the common case where the entire ``with: [...]`` expression
sits on a single line.  Multi-line ``with:`` blocks are not transformed (a
warning is printed instead so you can handle them manually).

Usage
-----
    # Dry-run (preview changes, no files modified)
    python3 fix_with_pairs.py /path/to/workspace --dry-run

    # Apply changes in-place
    python3 fix_with_pairs.py /path/to/workspace

    # Restrict to a specific file extension (default: .gaml)
    python3 fix_with_pairs.py /path/to/workspace --ext .gaml
"""

import argparse
import os
import re
import sys

# ---------------------------------------------------------------------------
# Regex
# ---------------------------------------------------------------------------

# Matches the full  with: [ ... ]  construct on a single line.
# Group 1 → everything before the opening bracket (i.e. "with:\s*")
# Group 2 → the content inside the brackets (may contain any chars except ])
# We use a non-greedy match so we stop at the first ']'.
_PATTERN = re.compile(r'(with:\s*)\[([^\]]*)\]')


def _transform_content(content: str) -> str:
    """Replace every ``::`` inside a ``with:`` list with ``:``.

    Parameters
    ----------
    content:
        The raw text between the ``[`` and ``]`` delimiters.

    Returns
    -------
    str
        The same text with all ``::`` occurrences replaced by ``:``.
    """
    return content.replace('::', ':')


def _replace_match(match: re.Match) -> str:
    """Build the replacement string for a single regex match.

    Parameters
    ----------
    match:
        A :class:`re.Match` object produced by ``_PATTERN``.

    Returns
    -------
    str
        The transformed ``with: (...)`` expression.
    """
    prefix = match.group(1)          # e.g. "with: "
    content = match.group(2)         # everything between [ and ]
    return f'{prefix}({_transform_content(content)})'


def _transform_line(line: str) -> str:
    """Return *line* with all matching ``with: [...]`` patterns replaced.

    Parameters
    ----------
    line:
        A single line of source text.

    Returns
    -------
    str
        The transformed line (unchanged if no pattern matched).
    """
    return _PATTERN.sub(_replace_match, line)


def _process_file(path: str, dry_run: bool) -> bool:
    """Process a single file.

    Parameters
    ----------
    path:
        Absolute path to the file to process.
    dry_run:
        When ``True`` the file is not written; changes are only printed.

    Returns
    -------
    bool
        ``True`` if at least one substitution was made (or would be made).
    """
    try:
        with open(path, 'r', encoding='utf-8', errors='replace') as fh:
            original_lines = fh.readlines()
    except OSError as exc:
        print(f'  [ERROR] Cannot read {path}: {exc}', file=sys.stderr)
        return False

    new_lines = [_transform_line(line) for line in original_lines]

    if new_lines == original_lines:
        return False

    print(f'  {"[dry-run] " if dry_run else ""}Updating: {path}')
    for i, (old, new) in enumerate(zip(original_lines, new_lines), start=1):
        if old != new:
            print(f'    line {i}:')
            print(f'      - {old.rstrip()}')
            print(f'      + {new.rstrip()}')

    if not dry_run:
        try:
            with open(path, 'w', encoding='utf-8') as fh:
                fh.writelines(new_lines)
        except OSError as exc:
            print(f'  [ERROR] Cannot write {path}: {exc}', file=sys.stderr)
            return False

    return True


def _iter_files(root: str, extensions: list):
    """Yield all files under *root* whose suffix is in *extensions*.

    Hidden directories (names starting with ``.``) are skipped.

    Parameters
    ----------
    root:
        The directory to walk recursively.
    extensions:
        List of file extensions to include (e.g. ``['.gaml']``).
    """
    for dirpath, dirnames, filenames in os.walk(root):
        dirnames[:] = [d for d in dirnames if not d.startswith('.')]
        for fname in filenames:
            if any(fname.endswith(ext) for ext in extensions):
                yield os.path.join(dirpath, fname)


def main(argv=None):
    """Entry point for the script.

    Parameters
    ----------
    argv:
        Argument list (defaults to ``sys.argv[1:]``).

    Returns
    -------
    int
        Exit code (0 = success).
    """
    parser = argparse.ArgumentParser(
        description=(
            'Replace "with: [key::val, ...]" with "with: (key:val, ...)" '
            'in GAML files.'
        )
    )
    parser.add_argument('root', help='Root directory to search recursively.')
    parser.add_argument('--dry-run', action='store_true',
                        help='Preview changes without modifying any file.')
    parser.add_argument('--ext', action='append', dest='extensions',
                        default=['.gaml'],
                        metavar='EXT',
                        help='File extension to process (default: .gaml). '
                             'Repeat for multiple extensions.')
    args = parser.parse_args(argv)

    changed = 0
    total = 0
    for filepath in _iter_files(args.root, args.extensions):
        total += 1
        if _process_file(filepath, args.dry_run):
            changed += 1

    action = 'Would update' if args.dry_run else 'Updated'
    print(f'\n{action} {changed} / {total} file(s).')
    return 0


if __name__ == '__main__':
    sys.exit(main())

File: gaml.grammar/src/gaml/grammar/transition/FixWithPairs.java
package gaml.grammar.transition;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * {@link IFileTransformer} that transforms {@code with:} argument lists in
 * GAML source files from the old bracket / double-colon pair syntax to the new
 * parenthesis / single-colon syntax.
 *
 * <p>Examples:</p>
 * <pre>
 *   with: [with_viz::true, step::1#s, name::"Simulation step=1s"]
 *       → with: (with_viz:true, step:1#s, name:"Simulation step=1s")
 *
 *   with: [agents::ag, values::[1,2,3]]
 *       → with: (agents:ag, values:[1,2,3])
 * </pre>
 *
 * <p>Multi-line {@code with: [...]} blocks are also handled correctly because
 * the transformation operates on the entire file text at once.</p>
 *
 * <h3>Rules applied</h3>
 * <ol>
 *   <li>The opening {@code [} immediately following {@code with:} (with optional
 *       whitespace, including newlines) is replaced by {@code (}.</li>
 *   <li>The matching closing {@code ]} (respecting nested brackets and quoted
 *       strings) is replaced by {@code )}.</li>
 *   <li>Every {@code ::} that sits directly inside the outermost list (i.e. not
 *       inside a nested {@code [...]} or a quoted string) is replaced by
 *       {@code :}.</li>
 * </ol>
 *
 * <p>Nested {@code [...]} expressions and their contents are left completely
 * untouched.</p>
 */
public class FixWithPairs implements IFileTransformer {

	/**
	 * Pattern that finds {@code with:} followed by optional whitespace (including
	 * newlines) immediately before an opening {@code [}. The look-ahead ensures
	 * the {@code [} is present but not consumed, so that {@link #findMatchingBracket}
	 * can locate it precisely via {@link Matcher#end()}.
	 */
	private static final Pattern WITH_PREFIX = Pattern.compile("with:\\s*(?=\\[)");

	/**
	 * Find the index of the {@code ]} that closes the {@code [} at position
	 * {@code start} inside {@code text}.
	 *
	 * <p>The search respects:</p>
	 * <ul>
	 *   <li>Nested {@code [...]} brackets (depth tracking).</li>
	 *   <li>Double-quoted strings (characters inside quotes are ignored).</li>
	 *   <li>Newlines (the search continues across line boundaries).</li>
	 * </ul>
	 *
	 * @param text
	 *            the full source text; never {@code null}
	 * @param start
	 *            the index of the opening {@code [} character
	 * @return the index of the matching {@code ]}, or {@code -1} if not found
	 */
	static int findMatchingBracket(final String text, final int start) {
		int depth = 0;
		boolean inString = false;
		for (int i = start; i < text.length(); i++) {
			final char ch = text.charAt(i);
			if (inString) {
				if (ch == '\\') {
					i++; // skip escaped character
				} else if (ch == '"') {
					inString = false;
				}
			} else {
				if (ch == '"') {
					inString = true;
				} else if (ch == '[') {
					depth++;
				} else if (ch == ']') {
					depth--;
					if (depth == 0) { return i; }
				}
			}
		}
		return -1;
	}

	/**
	 * Replace every top-level {@code ::} inside {@code content} with {@code :}.
	 *
	 * <p>Characters inside nested {@code [...]} brackets or double-quoted strings
	 * are left completely untouched.</p>
	 *
	 * @param content
	 *            the raw text between the outermost {@code [} and {@code ]}
	 *            delimiters (not including those delimiters themselves);
	 *            never {@code null}
	 * @return the transformed content; never {@code null}
	 */
	static String transformContent(final String content) {
		final StringBuilder result = new StringBuilder(content.length());
		int depth = 0;
		boolean inString = false;
		for (int i = 0; i < content.length(); i++) {
			final char ch = content.charAt(i);
			if (inString) {
				if (ch == '\\') {
					result.append(ch);
					result.append(content.charAt(i + 1));
					i++;
					continue;
				}
				if (ch == '"') { inString = false; }
				result.append(ch);
			} else {
				if (ch == '"') {
					inString = true;
					result.append(ch);
				} else if (ch == '[') {
					depth++;
					result.append(ch);
				} else if (ch == ']') {
					depth--;
					result.append(ch);
				} else if (ch == ':' && depth == 0
						&& i + 1 < content.length() && content.charAt(i + 1) == ':') {
					// Top-level "::" → replace with single ":"
					result.append(':');
					i++; // skip the second ':'
				} else {
					result.append(ch);
				}
			}
		}
		return result.toString();
	}

	/**
	 * Apply the {@code with: [...]} → {@code with: (...)} transformation to the
	 * given source content.
	 *
	 * <p>Uses a bracket-aware parser so that nested {@code [...]} inside the list
	 * are preserved intact, and multi-line {@code with: [...]} blocks are handled
	 * correctly.</p>
	 *
	 * @param content
	 *            the full text of a GAML source file; never {@code null}
	 * @return the transformed text; the original string is returned when no
	 *         match is found
	 */
	@Override
	public String transform(final String content) {
		final StringBuilder result = new StringBuilder(content.length());
		int pos = 0;
		final Matcher m = WITH_PREFIX.matcher(content);
		boolean changed = false;

		while (m.find()) {
			// Skip matches that fall inside an already-consumed region
			if (m.start() < pos) { continue; }

			final int bracketStart = m.end(); // index of the '['
			final int bracketEnd = findMatchingBracket(content, bracketStart);
			if (bracketEnd == -1) { continue; } // no matching ']' — skip

			// Append everything up to the '['
			result.append(content, pos, bracketStart);
			// Opening '(' instead of '['
			result.append('(');
			// Transformed inner content
			result.append(transformContent(content.substring(bracketStart + 1, bracketEnd)));
			// Closing ')' instead of ']'
			result.append(')');
			pos = bracketEnd + 1;
			changed = true;
		}

		if (!changed) { return content; }
		result.append(content, pos, content.length());
		return result.toString();
	}
}
